### ISSUE TITLE
Sidebar Navigator UI Corruption [Investigation]

### ISSUE THREAD

**Issue by rat-moonshine** [Moonshine-IDE/Moonshine-IDE#1046]
From @JustinProminic 

-----------------------------------------

I was able to corrupt the left hand navigator UI.

I did this easily with a project that has many subfolders.

By opening and collapsing, I ended up first with one blank line that should not exist, and then this giant white area of undefined space.

![image](https://user-images.githubusercontent.com/24370870/172118321-f037cde9-0092-4cc3-b389-7c2157f111a9.png)

I had to shut down Moonshine and restart it.

When I did that, one of the projects was gone from the workspace.   The very bottom project was never closed by me.  It simply didn't exist anymore in the workspace.

You can see the nesting depth of the folder I was opening and closing (along with the com one) to reproduce the above UI failure.

![image](https://user-images.githubusercontent.com/24370870/172118449-9cd4f2c0-3398-405d-b370-bfb167ca7552.png)

**Comment by Igazine** [comment]
I've created a project to test with 10s of thousands of subdirs and files, and indeed I can reproduce this issue by opening and closing the project.

**Note: So far it does not occur in build 2601, so it looks fine after the adjustment, but I'll test it further.**

I've also noticed that if I move the resizer to the right after the problem appears, all of the file and dir items are actually in the right place, but they're covered by a white rectangle (please see screenshot), so it seems it's indeed an issue with Flex component redraw (or something similar)

I've also noticed two issues regarding the tree view component (please ignore them if they were already reported):

* Single click (on the arrow icon) and double click (on a file/directory) behave differently. Double-click only opens the directory if it has been previously opened by clicking the arrow. Please take a look at the highlighted item in the screenshot, it should be opened showing its contents, but since I've opened it with a double-click, only the icon has changed.
* If I double-click the project's name it opens the project with all of the previously opened subdirectories opened, but if I click on the arrow, every subdir is closed. Might be the intentional behavior, I'm not sure.

<img width="662" alt="image" src="https://user-images.githubusercontent.com/2218406/172373896-a40beb02-a369-4427-b865-d4057ecd15ab.png">

**Comment by rat-moonshine** [comment]
Flex Tree menu in our experience we found always buggy. To overcome many such bugs/limitations the current `TreeView` component is full of hacks and nasty codes. At this point this even become problematic to remember which hacks was doing what, in that component.

**Comment by JoelProminic** [comment]
I was finally able to reproduce this bug on an old build by expanding and collapsing directories by double-clicking them.  I haven't seen this trigger when I use the twistie instead.  The procedure is something like this:
1. Open MoonshineDESKTOPevolved (or restart Moonshine with the project already open)
2. Expand src with the twistie
3. Expand or collapse notes under src
4. Collapse and expand src *by double-clicking*
5. If the bug doesn't trigger, repeat from step 3.  It takes a while to trigger for me, and I haven't been able to trigger it consistently.

I don't normally use the double-click method for this.  When I tried using this today, I noticed some odd behaviors related to this.  These are probably issues that can wait until we migrate to Haxe, though.
- Double-click doesn't work if I have not opened the directory before
- Double-click doesn't work on the root project node
- Double-click expands the nested expanded directory, but the twistie only reopens the immediate children of the node

I haven't been able to reproduce this with the latest macOS build yet.  I'll plan to do some tests on Windows as well.
